shell/server/core.js

```diff
@@ -112,33 +116,35 @@ NotificationHandle.prototype.close = function () {
   });
 };
 
-NotificationHandle.prototype.save = function (params) {
-  var self = this;
+saveFrontendRef = function (frontendRef, owner) {
   return inMeteor(function () {
-    var sturdyRef = new Buffer(Random.id(20));
+    var sturdyRef = new Buffer(generateSturdyRef());
     var hashedSturdyRef = hashSturdyRef(sturdyRef);
     ApiTokens.insert({
       _id: hashedSturdyRef,
-      frontendRef: {
-        notificationHandle: self.notificationId
-      },
-      owner: params.sealFor,
+      frontendRef: frontendRef,
+      owner: owner,
+      created: new Date()
     });
-    self.saved = true;
     return {sturdyRef: sturdyRef};
   });
 };
 
-function restoreInternal(sturdyRef, ownerPattern) {
-  // Restores `sturdyRef`, checking first that its owner matches `ownerPattern`.
+NotificationHandle.prototype.save = function (params) {
+  return saveFrontendRef({notificationHandle: this.notificationId}, params.sealFor);
+};
 
+restoreInternal = function (sturdyRef, ownerPattern, requiredPermissions) {
```

Comment:
I don't think that the MembraneRequirement represented by `requiredPermissions` should be added to the existing token. I think that `SandstormCore.restore(token, requiredPermissions)` should call `restoreInternal()` and then call `SystemPersistent.addRequirements()` on the result. If the restored capability eventually gets re-saved, then that's when the new MembraneRequirement should get persisted.